espnet/nets/time_sync_beam_search.py

```diff
@@ -0,0 +1,228 @@
+"""
+Time Synchronous One-Pass Beam Search -
+Implements joint CTC/attention decoding where
+hypotheses are expanded along the time (input) axis,
+as described in https://arxiv.org/abs/2210.05200
+References: https://arxiv.org/abs/1408.2873 for CTC beam search
+Author: Brian Yan
+"""
+
+from dataclasses import dataclass
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Union
+from typing import Tuple
+
+import numpy as np
+import torch
+from typeguard import check_argument_types
+
+from collections import defaultdict
+import logging
+
+from espnet.nets.beam_search import Hypothesis
+from espnet.nets.scorer_interface import ScorerInterface
+
+
+@dataclass
+class CacheItem:
+    """For caching attentional decoder and LM states."""
+
+    state: Any
+    scores: Any
+    log_sum: float
+
+
+class TimeSyncBeamSearch(torch.nn.Module):
+    """Time synchronous beam search algorithm"""
+
+    def __init__(
+        self,
+        sos: int,
+        beam_size: int,
+        ctc: torch.nn.Module,
+        token_list=dict,
+        pre_beam_ratio: float = 1.5,
+        decoder: ScorerInterface = None,
+        lm: ScorerInterface = None,
+        ctc_weight: float = 1.0,
+        lm_weight: float = 0.0,
+        penalty: float = 1.0,
+        blank: int = 0,
+        force_lid: bool = False,
+        temp: float = 1.0,
+    ):
+        """
+        beam_size: num hyps
+        sos: sos index
+        ctc: CTC module
+        pre_beam_ratio: pre_beam_ratio * beam_size = pre_beam
```

Comment:
It is better to have more explanations of pre_beam

Reply:
done